File: data structures and algorithms/java/06_110.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

public class FloydWarshall {
   // Avoid overflow when adding distances
   private static final int INF = Integer.MAX_VALUE / 2;
   
   /**
    * Main method to demonstrate the algorithm
    */
   public static void main(String[] args) {
      // Example graph represented as adjacency matrix
      int[][] graph = {
         {0, 5, INF, 10},
         {INF, 0, 3, INF},
         {INF, INF, 0, 1},
         {INF, INF, INF, 0}
      };
      
      // Run Floyd-Warshall algorithm
      FloydWarshallResult result = floydWarshall(graph);
      
      // Print all-pairs shortest paths
      System.out.println("Shortest distances:");
      for (int[] row : result.distances) {
         System.out.println(Arrays.toString(row));
      }
      
      // Find path from vertex 0 to 3
      List<Integer> path = getPath(0, 3, result.nextVertex);
      if (!path.isEmpty()) {
         System.out.println("Path from 0 to 3: " + path);
      } else {
         System.out.println("No path exists from 0 to 3");
      }
   }
   
   /**
    * Result container for the algorithm
    */
   static class FloydWarshallResult {
      int[][] distances;
      Integer[][] nextVertex;
      
      FloydWarshallResult(int[][] distances, 
            Integer[][] nextVertex) {
         this.distances = distances;
         this.nextVertex = nextVertex;
      }
   }
   
   /**
    * Floyd-Warshall all-pairs shortest path algorithm
    */
   public static FloydWarshallResult floydWarshall(
         int[][] graph) {
      int n = graph.length;
      int[][] dist = new int[n][n];
      Integer[][] next = new Integer[n][n];
      
      // Initialize distance and next matrices
      for (int i = 0; i < n; i++) {
         for (int j = 0; j < n; j++) {
            dist[i][j] = graph[i][j];
            if (graph[i][j] != INF && i != j) {
               next[i][j] = j;
            }
         }
      }
      
      // Main algorithm - consider all vertices as intermediates
      for (int k = 0; k < n; k++) {
         for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
               if (dist[i][k] != INF && dist[k][j] != INF) {
                  // If path through k is shorter
                  if (dist[i][j] > dist[i][k] + dist[k][j]) {
                     dist[i][j] = dist[i][k] + dist[k][j];
                     next[i][j] = next[i][k];
                  }
               }
            }
         }
      }
      
      return new FloydWarshallResult(dist, next);
   }
   
   /**
    * Reconstruct path from start to end using next matrix
    */
   public static List<Integer> getPath(int start, int end,
         Integer[][] next) {
      List<Integer> path = new ArrayList<>();
      
      // No path exists
      if (next[start][end] == null) {
         return path;
      }
      
      // Reconstruct path
      path.add(start);
      while (start != end) {
         start = next[start][end];
         path.add(start);
      }
      
      return path;
   }
}

File: data structures and algorithms/cs/06_110.cs
using System;
using System.Collections.Generic;

public class FloydWarshall
{
   private static readonly int INF = int.MaxValue / 2;

   public static void Main(string[] args)
   {
      int[][] graph = {
            new int[] {0, 5, INF, 10},
            new int[] {INF, 0, 3, INF},
            new int[] {INF, INF, 0, 1},
            new int[] {INF, INF, INF, 0}
        };

      var result = FloydWarshallAlgorithm(graph);

      Console.WriteLine("Shortest distances:");
      foreach (var row in result.Distances)
      {
         Console.WriteLine(string.Join(", ", row));
      }

      var path = GetPath(0, 3, result.NextVertex);
      if (path.Count > 0)
      {
         Console.WriteLine("Path from 0 to 3: " +
            string.Join(" -> ", path));
      }
   }

   public class FloydWarshallResult
   {
      public int[][] Distances { get; set; }
      public int?[][] NextVertex { get; set; }
   }

   public static FloydWarshallResult
      FloydWarshallAlgorithm(int[][] graph)
   {
      int n = graph.Length;
      int[][] dist = new int[n][];
      int?[][] next = new int?[n][];

      for (int i = 0; i < n; i++)
      {
         dist[i] = new int[n];
         next[i] = new int?[n];
         for (int j = 0; j < n; j++)
         {
            dist[i][j] = graph[i][j];
            if (graph[i][j] != INF && i != j)
            {
               next[i][j] = j;
            }
         }
      }

      for (int k = 0; k < n; k++)
      {
         for (int i = 0; i < n; i++)
         {
            for (int j = 0; j < n; j++)
            {
               if (dist[i][k] + dist[k][j] < dist[i][j])
               {
                  dist[i][j] = dist[i][k] + dist[k][j];
                  next[i][j] = next[i][k];
               }
            }
         }
      }

      return new FloydWarshallResult
      {
         Distances = dist,
         NextVertex = next
      };
   }

   public static List<int> GetPath(int start,
      int end, int?[][] next)
   {
      if (next[start][end] == null)
      {
         return new List<int>();
      }

      List<int> path = new List<int> { start };
      while (start != end)
      {
         start = next[start][end].Value;
         path.Add(start);
      }

      return path;
   }
}
